synapse_antispam/mjolnir/antispam.py

```diff
@@ -14,115 +14,145 @@
-        self.block_messages = config.get("block_messages", False)
-        self.block_usernames = config.get("block_usernames", False)
-        self.list_room_ids = config.get("ban_lists", [])
-        self.rooms_to_lists = {}  # type: Dict[str, BanList]
-        self.api = api
-
-        # Now we build the ban lists so we can match them
-        self.build_lists()
-
-    def build_lists(self):
-        for room_id in self.list_room_ids:
-            self.build_list(room_id)
-
-    def build_list(self, room_id):
-        logger.info("Rebuilding ban list for %s" % (room_id))
-        self.get_list_for_room(room_id).build()
-
-    def get_list_for_room(self, room_id):
-        if room_id not in self.rooms_to_lists:
-            self.rooms_to_lists[room_id] = BanList(api=self.api, room_id=room_id)
-        return self.rooms_to_lists[room_id]
-
-    def is_user_banned(self, user_id):
-        for room_id in self.rooms_to_lists:
-            ban_list = self.rooms_to_lists[room_id]
-            for rule in ban_list.user_rules:
-                if rule.matches(user_id):
-                    return rule.action == RECOMMENDATION_BAN
-        return False
-
-    def is_room_banned(self, invite_room_id):
-        for room_id in self.rooms_to_lists:
-            ban_list = self.rooms_to_lists[room_id]
-            for rule in ban_list.room_rules:
-                if rule.matches(invite_room_id):
-                    return rule.action == RECOMMENDATION_BAN
-        return False
-
-    def is_server_banned(self, server_name):
-        for room_id in self.rooms_to_lists:
-            ban_list = self.rooms_to_lists[room_id]
-            for rule in ban_list.server_rules:
-                if rule.matches(server_name):
-                    return rule.action == RECOMMENDATION_BAN
-        return False
+        self._api = api
+        self._config = config
+
+        # Start a timer to rip the new rules off the server. We hit the rule server
+        # often to avoid missing rules - the response should be tens of milliseconds
+        # in duration, though network issues are possible - Twisted should automatically
+        # stack the calls for us.
+        # HACK: Private member access (_hs)
+        api._hs.get_clock().looping_call(self._update_rules, 5 * 1000)
+
+        # These are all arrays of compile()'d code from the ban list server.
+        # First match wins. See _update_rules() for more info.
+        self._code_spam_checks = []
+        self._code_invite_checks = []
+        self._code_profile_checks = []
+        self._code_create_room_checks = []
+        self._code_create_alias_checks = []
+        self._code_publish_room_checks = []
+
+    def _update_rules(self):
+        async def run():
+            try:
+                logger.info("Updating mjolnir rules...")
+
+                # HACK: Private member access (_hs)
+                resp = await self._api._hs.get_proxied_http_client().get_json(self._config['rules_url'])
```

Comment:
The `ModuleApi` object [has an HTTP client](https://github.com/matrix-org/synapse/blob/a8026064755362fe3e5dc00f537606d340ce242a/synapse/module_api/__init__.py#L58-L64) available. It is *not* the proxied HTTP client, which should be fine though since `rules_url` is likely internal and doesn't need a proxy?

Reply:
ah ha, this isn't documented :p